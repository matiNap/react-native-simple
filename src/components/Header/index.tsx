import React from 'react';
import { StyleSheet, View } from 'react-native';
import metrics from '../../theme/metrics';
import useTheme from '../ThemeProvider/useTheme';
import { getThemeColor, getShadow, getProperty } from '../../helpers';
import Props from '../../types/Header';

const DEFAULT_HEIGHT = metrics.screen.height * 0.12;
const FIXED_STYLE = {
    position: 'absolute',
    top: 0,
    left: 0,
};
export default ({ children, ...props }: Props) => {
    const theme = useTheme();
    const { currentPalette, Header } = theme;
    const translucent = getProperty([Header?.translucent, props.translucent]);
    const fixed = getProperty([Header?.fixed, props.fixed]);
    const style = getProperty([Header?.style, props.style]);
    const backgroundColor = getProperty([
        currentPalette.primary,
        Header?.backgroundColor,
        getThemeColor(props.backgroundColor, currentPalette),
    ]);
    const radiusProperty = getProperty([
        Header?.borderRadius,
        props.borderRadius,
    ]);
    const borderRadius = radiusProperty ? radiusProperty : 0;
    const height = getProperty([DEFAULT_HEIGHT, Header?.height, props.height]);
    const paddingTop = translucent ? metrics.statusBarHeight + 10 : 0;
    const shadow = getProperty([Header?.shadow, getShadow(props.shadow)]);
    const fixedStyle = fixed ? FIXED_STYLE : {};

    return (
        <View
            style={StyleSheet.flatten([
                styles.container,
                {
                    backgroundColor,
                    height,
                    paddingTop,
                    borderBottomLeftRadius: borderRadius,
                    borderBottomRightRadius: borderRadius,
                    ...shadow,
                    ...fixedStyle,
                },
                style,
            ])}
        >
            {children}
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        width: '100%',
        zIndex: 160,
        flexDirection: 'row',
        justifyContent: 'space-between',
        alignItems: 'center',
        paddingHorizontal: 10,
    },
});
