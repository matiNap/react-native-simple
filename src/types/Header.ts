import { StyleProp, ViewStyle } from 'react-native';
import { ReactNode } from 'react';
import { SimpleShadow, SimpleBackgroundColor } from '.';

export default interface HeaderPorps {
    height?: number;
    backgroundColor?: SimpleBackgroundColor;
    shadow?: SimpleShadow;
    translucent?: boolean | undefined;
    children?: ReactNode | ReactNode[];
    borderRadius?: number;
    style?: StyleProp<ViewStyle>;
    fixed?: boolean | undefined;
}
