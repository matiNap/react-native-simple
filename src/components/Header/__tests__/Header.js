import React from 'react';
import { shallow } from 'enzyme';
import { describe, expect, it } from '@jest/globals';
import Header from '../index';
import toJson from 'enzyme-to-json';
import { Text } from 'react-native';

describe('Header component', () => {
    it('Should render', () => {
        const wrapper = shallow(<Header />);

        expect(wrapper.length).toBe(1);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('Should render child', () => {
        const wrapper = shallow(
            <Header>
                <Text>A</Text>
            </Header>,
        );

        expect(wrapper.find(Text).length).toBe(1);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('Should render multiple childs', () => {
        const wrapper = shallow(
            <Header>
                <Text>A</Text>
                <Text>B</Text>
            </Header>,
        );

        expect(wrapper.find(Text).length).toBe(2);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('Should render with style', () => {
        const wrapper = shallow(<Header style={{ width: '80%' }} />);

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('Should render with shadow', () => {
        const wrapper = shallow(<Header shadow="default" />);

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('Should render fixed', () => {
        const wrapper = shallow(<Header shadow="default" fixed />);

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('Should render rounded borders', () => {
        const wrapper = shallow(<Header borderRadius={20} />);

        expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('Should render with new backround color', () => {
        const wrapper = shallow(
            <Header borderRadius={20} backgroundColor="#313131" />,
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
