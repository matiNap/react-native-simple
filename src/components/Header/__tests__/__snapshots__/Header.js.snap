// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component Should render 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ededed",
      "borderBottomLeftRadius": 0,
      "borderBottomRightRadius": 0,
      "flexDirection": "row",
      "height": 160.07999999999998,
      "justifyContent": "space-between",
      "paddingHorizontal": 10,
      "paddingTop": 0,
      "width": "100%",
      "zIndex": 160,
    }
  }
/>
`;

exports[`Header component Should render child 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ededed",
      "borderBottomLeftRadius": 0,
      "borderBottomRightRadius": 0,
      "flexDirection": "row",
      "height": 160.07999999999998,
      "justifyContent": "space-between",
      "paddingHorizontal": 10,
      "paddingTop": 0,
      "width": "100%",
      "zIndex": 160,
    }
  }
>
  <Text>
    A
  </Text>
</View>
`;

exports[`Header component Should render fixed 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ededed",
      "borderBottomLeftRadius": 0,
      "borderBottomRightRadius": 0,
      "elevation": 7,
      "flexDirection": "row",
      "height": 160.07999999999998,
      "justifyContent": "space-between",
      "left": 0,
      "paddingHorizontal": 10,
      "paddingTop": 0,
      "position": "absolute",
      "shadowColor": "#000",
      "shadowOffset": Object {
        "height": 3,
        "width": 0,
      },
      "shadowOpacity": 0.29,
      "shadowRadius": 4.65,
      "top": 0,
      "width": "100%",
      "zIndex": 160,
    }
  }
/>
`;

exports[`Header component Should render multiple childs 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ededed",
      "borderBottomLeftRadius": 0,
      "borderBottomRightRadius": 0,
      "flexDirection": "row",
      "height": 160.07999999999998,
      "justifyContent": "space-between",
      "paddingHorizontal": 10,
      "paddingTop": 0,
      "width": "100%",
      "zIndex": 160,
    }
  }
>
  <Text>
    A
  </Text>
  <Text>
    B
  </Text>
</View>
`;

exports[`Header component Should render rounded borders 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ededed",
      "borderBottomLeftRadius": 20,
      "borderBottomRightRadius": 20,
      "flexDirection": "row",
      "height": 160.07999999999998,
      "justifyContent": "space-between",
      "paddingHorizontal": 10,
      "paddingTop": 0,
      "width": "100%",
      "zIndex": 160,
    }
  }
/>
`;

exports[`Header component Should render with new backround color 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#313131",
      "borderBottomLeftRadius": 20,
      "borderBottomRightRadius": 20,
      "flexDirection": "row",
      "height": 160.07999999999998,
      "justifyContent": "space-between",
      "paddingHorizontal": 10,
      "paddingTop": 0,
      "width": "100%",
      "zIndex": 160,
    }
  }
/>
`;

exports[`Header component Should render with shadow 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ededed",
      "borderBottomLeftRadius": 0,
      "borderBottomRightRadius": 0,
      "elevation": 7,
      "flexDirection": "row",
      "height": 160.07999999999998,
      "justifyContent": "space-between",
      "paddingHorizontal": 10,
      "paddingTop": 0,
      "shadowColor": "#000",
      "shadowOffset": Object {
        "height": 3,
        "width": 0,
      },
      "shadowOpacity": 0.29,
      "shadowRadius": 4.65,
      "width": "100%",
      "zIndex": 160,
    }
  }
/>
`;

exports[`Header component Should render with style 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ededed",
      "borderBottomLeftRadius": 0,
      "borderBottomRightRadius": 0,
      "flexDirection": "row",
      "height": 160.07999999999998,
      "justifyContent": "space-between",
      "paddingHorizontal": 10,
      "paddingTop": 0,
      "width": "80%",
      "zIndex": 160,
    }
  }
/>
`;
